### ISSUE TITLE
Some wired pink-green artifact

### ISSUE THREAD

**Issue by YotamWolf** [TEXTurePaper/TEXTurePaper#8]
Hi thanks for the amazing project, the best I've seen so far.

On some of the 3D models I'm trying I receive som weird pink-green artifact like salt and pepper noise, any idea why is that
![final_obj_7bc0a1e4-ac51-11ed-a5e4-06e882d545ae](https://user-images.githubusercontent.com/38569619/218726043-64e3ff1f-026c-4d64-8a83-942990bef4dc.png)
 happening? 
<img width="473" alt="Screenshot 2023-02-14 at 13 19 47" src="https://user-images.githubusercontent.com/38569619/218726063-b73dda7f-cccd-4c12-9bc7-3133b7daaa20.png">

**Comment by AmanKishore** [comment]
I'm getting a similar issue: 
<img width="512" alt="image" src="https://user-images.githubusercontent.com/10477380/220823371-a6e65c3c-9a0b-4cd9-b832-948c91124ef3.png">

**Comment by pengwangucla** [comment]
+1

**Comment by YuliangXiu** [comment]
+1

![image](https://user-images.githubusercontent.com/7944350/229275969-1d1edc3d-fdff-4a6a-a02e-b3cce47f5362.png)

**Comment by qzq2514** [comment]
Differentiable rendering has reconstruction error under the api of kaolin，this is the main reason to cause this green artifact. Pink artifact is caused by its inaccurate method to calculate trimap.

**Comment by pengwangucla** [comment]
> Differentiable rendering has reconstruction error under the api of kaolin，this is the main reason to cause this green artifact. Pink artifact is caused by its inaccurate method to calculate trimap.

any way to fix these issue ?  e.g. for the first issue, should we use pytorch3d differential renderring or there is a corrected way.

**Comment by qzq2514** [comment]
> > Differentiable rendering has reconstruction error under the api of kaolin，this is the main reason to cause this green artifact. Pink artifact is caused by its inaccurate method to calculate trimap.
> 
> any way to fix these issue ? e.g. for the first issue, should we use pytorch3d differential renderring or there is a corrected way.

I haven't try pytorch3d differential renderring, but I think this error may still exist for any texture reconstruction under differentiable rendering, but you can try to map the view image to texture space directly, just like the inverse of kaolin's texture_mapping

**Comment by YuzhiChen001** [comment]
> Differentiable rendering has reconstruction error under the api of kaolin，this is the main reason to cause this green artifact. Pink artifact is caused by its inaccurate method to calculate trimap.

"inaccurate method to calculate trimap",I want to know why the author's method is inaccurate?How do you think it should be calculated?thanks.

**Comment by qzq2514** [comment]
> method
author's method to get uncolored mask is to calculate difference between textured uvmap and base color, this will be mostly affected by the color of texture. The correct method is to calculate the accurate by reverse of rasterizing.

**Comment by qzq2514** [comment]
> > Differentiable rendering has reconstruction error under the api of kaolin，this is the main reason to cause this green artifact. Pink artifact is caused by its inaccurate method to calculate trimap.
> 
> "inaccurate method to calculate trimap",I want to know why the author's method is inaccurate?How do you think it should be calculated?thanks.

author's method to get uncolored mask is to calculate difference between textured uvmap and base color, this will be mostly affected by the color of texture. The correct method is to use  reverse process of rasterizing.

**Comment by smha-Promedius** [comment]
This problem occurs because the mesh color is not optimized in trainer.py. However, this is not a fundamental problem and is closely related to the normal parameters from the mesh(=`self.mesh_model.get_params()` in `Adam` optimizer) used and the `train/eval_grid_size` used for learning.

**Comment by qzq2514** [comment]
> > > Differentiable rendering has reconstruction error under the api of kaolin，this is the main reason to cause this green artifact. Pink artifact is caused by its inaccurate method to calculate trimap.
> > 
> > 
> > "inaccurate method to calculate trimap",I want to know why the author's method is inaccurate?How do you think it should be calculated?thanks.
> 
> author's method to get uncolored mask is to calculate difference between textured uvmap and base color, this will be mostly affected by the color of texture. The correct method is to use reverse process of rasterizing.

We have in part solved this problem, you can find this in https://github.com/OpenTexture/Paint3D
